Add blog post for Android tracker 1.4.0 (#1328)
snowplow/snowplowanalytics.com: _posts/2020-03-05-snowplow-android-tracker-1.4.0-released/index.md

|`basisForProcessing` | GDPR Basis for processing | Yes           | Enum String   |
|     `documentId` | ID of a GDPR basis document     | No           | String   |
|        `documentVersion` | Version of the document        | No            | String   |
| `documentDescription` | Description of the document | No            | String   |

The required basisForProcessing accepts only the following literals: `consent`, `contract`, `legal_obligation`, `vital_interests`, `public_task`, `legitimate_interests` - in accordance with the [five legal bases for processing](https://ico.org.uk/for-organisations/guide-to-data-protection/guide-to-the-general-data-protection-regulation-gdpr/lawful-basis-for-processing/)

The GDPR context is enabled by calling the `gdprContext` method of the tracker builder or by calling the `enableGdprContext` method once the tracker has been initialised.

It is called as follows:

{% highlight java %}
Tracker.instance().enableGdprContext(
    Gdpr.Basis.CONSENT,
    "someId",
    "0.1.0",
    "a demo document description"
);
{% endhighlight %}

<h2 id="diagnostic">2. Diagnostic feature for internal error tracking</h2>

> The tracker is constantly tested with different platforms and versions but there are rare situations where an error can be reported. The tracker manages errors internally, avoiding crashes of the app and assuring that the events are not lost. However, it can now report errors to a Snowplow collector as `diagnostic_error` events.

Reviewer: ```suggestion
The tracker is constantly tested with different platforms and versions but there are rare situations where an error can be reported. The tracker manages errors internally, avoiding crashes of the app and assuring that the events are not lost. Now, the tracker can also report errors to a Snowplow collector as `diagnostic_error` events.
```

To activate this feature you only need to enable `trackerDiagnostic` ([#343][343]):

{% highlight java %}
TrackerBuilder trackerBuilder =
    new TrackerBuilder(emitter, namespace, appId, appContext)
        .trackerDiagnostic(true)
        ...
        .build();
Tracker.init(trackerBuilder);
{% endhighlight %}

> It will automatically report any internal tracker error to the configured Snowplow collector. These will be fundamental for the troubleshooting unexpected behaviour of the tracker.

Reviewer: ```suggestion
It will automatically report any internal tracker errors to the configured Snowplow collector. These will be fundamental for troubleshooting unexpected behaviour of the tracker.
```

For Snowplow Insights customers, the sequence of tracker diagnostic events can be sent to the Snowplow support team for analysis in order to spot tracker issues or the source of unexpected behaviour.

“unstruct_event_com_snowplowanalytics_snowplow_diagnostic_error_1”: {
      “className”: “Tracker”,
      “message”: “pool-1-thread-2|Error tracker message”
}

> <h2 id="iab">3. Fix for IAB enrichment filtering</h2>

Reviewer 2: I would avoid using 'filtering' at all - the enrichment itself doesn't filter, just adds a context.

> As reported [here](https://discourse.snowplowanalytics.com/t/warning-iab-enrichment-treats-android-tracker-events-as-spider-generated/2482) the IAB enrichment incorrectly treats events sent by the Android tracker as spider-generated. This is due to the fact that the default User Agent used by the Android tracker, which is the one of the underlying library used to make HTTP calls (okHttp), is part of the IAB blacklist.

Reviewer: This sentence is a bit confusing: 
This is due to the fact that the default User Agent used by the Android tracker, which is the one of the underlying library used to make HTTP calls (okHttp), is part of the IAB blacklist.

Does it mean: 
This is due to the fact that the underlying library used to make HTTP calls (okHttp) for the default User Agent used by the Android tracker is part of the IAB blacklist.

Author: I think we can omit all these details. The real meaning is:
The Android Tracker uses okHttp library. The default User-Agent of that library is in the IAB blacklist. Hence the Android tracker is seen as spider-generator.

What if I just say _This is due to the fact that the default User Agent used by the Android tracker is part of the IAB blacklist due to the underlying library `okHttp`._ ?

Reviewer 2: My 2c - This is actually the clearest explanation (just changing to past tense to denote it's fixed now):

`The Android Tracker uses okHttp library. The default User-Agent of that library is in the IAB blacklist. Hence the Android tracker was seen as spider-generator.`

Reviewer 2: Or: 

`The Android Tracker uses the okHttp library. The default User-Agent of that library is in the IAB blacklist, which caused the Android tracker to be seen as spider-generator.`

> The new Android tracker set a different default User Agent that is not filtered out by the IAB enrichment ([#359][359]). It's always possible to customize the User Agent used by the Android tracker setting the User Agent of the tracker's Subject: `setUserAgent`.

Reviewer: ```suggestion
The new Android tracker sets a different default User Agent that is not filtered out by the IAB enrichment ([#359][359]). It's always possible to customize the User Agent used by the Android tracker by setting the User Agent of the tracker's Subject: `setUserAgent`.
```

        ...
        .build();
TrackerBuilder trackerBuilder =
    new TrackerBuilder(emitter, namespace, appId, appContext)
        ...
        .subject(subject)
        ...
        .build();
Tracker.init(trackerBuilder);
{% endhighlight %}

<h2 id="custompath">4. Make POST path of Emitter configurable</h2>

This is a feature already introduced in the Obj-C tracker as of version 1.1.0, and is now available in the Android tracker ([#319][319]).
With the new version there is a new `customPostPath` parameter that makes the Emitter configurable.
It allows specifying a custom path for POST requests, this is useful when the user wants to use a proxy for receiving the events.

{% highlight java %}
Emitter emitter =
    new Emitter.EmitterBuilder(uri, context)
        ...
        .method(HttpMethod.POST)
        .customPostPath("com.acme.company/tpx")
        ...
        .build();
TrackerBuilder trackerBuilder =
    new TrackerBuilder(emitter, namespace, appId, appContext)
        ...
        .build();
Tracker.init(trackerBuilder);
{% endhighlight %}

<h2 id="updates">5. Updates and bug fixes</h2>

- Bump target Android API to 29 ([#357][357])

> We updated the library and the demo app to be fully compatible with Android 10 (API 29). It doesn't introduce any incompatibility with old versions and minimum Android OS version supported still Android 4.0 (API 14).

Reviewer: ```suggestion
We updated the library and the demo app to be fully compatible with Android 10 (API 29). It doesn't introduce any incompatibility with old versions and the minimum Android OS version supported is still Android 4.0 (API 14).
```

---
layout: post
title: "Snowplow Android Tracker 1.4.0 released"
title-short: Snowplow Android Tracker 1.4.0
tags: [snowplow, android, tracker]
author: Alex Benini
category: Releases
permalink: /blog/2020/03/05/snowplow-android-tracker-1.4.0-released/
discourse: false
---

We are pleased to announce a new release of the [Snowplow Android Tracker][android-tracker].

> [Version 1.4.0][1.4.0-tag] introduces support for GDPR contexts and tracker diagnostic useful for the troubleshooting in case of unexpected tracker behaviour.

Reviewer 3: ```suggestion
[Version 1.4.0][1.4.0-tag] introduces support for GDPR contexts and tracker diagnostics, useful for troubleshooting cases of unexpected tracker behaviour.
```

> Also included in this release, a fix to stop IAB enrichment from considering Android tracker events as spider-generated events.

Reviewer 3: ```suggestion
Also included in this release, a fix to stop the IAB enrichment from considering Android tracker events as spider-generated events.
```

|`basisForProcessing` | GDPR Basis for processing | Yes           | Enum String   |
|     `documentId` | ID of a GDPR basis document     | No           | String   |
|        `documentVersion` | Version of the document        | No            | String   |
| `documentDescription` | Description of the document | No            | String   |

The required basisForProcessing accepts only the following literals: `consent`, `contract`, `legal_obligation`, `vital_interests`, `public_task`, `legitimate_interests` - in accordance with the [five legal bases for processing](https://ico.org.uk/for-organisations/guide-to-data-protection/guide-to-the-general-data-protection-regulation-gdpr/lawful-basis-for-processing/)

The GDPR context is enabled by calling the `gdprContext` method of the tracker builder or by calling the `enableGdprContext` method once the tracker has been initialised.

It is called as follows:

{% highlight java %}
Tracker.instance().enableGdprContext(
    Gdpr.Basis.CONSENT,
    "someId",
    "0.1.0",
    "a demo document description"
);
{% endhighlight %}

<h2 id="diagnostic">2. Diagnostic feature for internal error tracking</h2>

> The tracker is constantly tested with different platforms and versions but there are rare situations where an error can be reported. The tracker can manage the situation avoiding crashes of the app and assuring that the events are not lost. However, it can report errors to the collector as `diagnostic_error` events.

Reviewer 3: ```suggestion
The tracker is constantly tested with different platforms and versions but there are rare situations where an error can be reported. The tracker manages errors internally, avoiding crashes of the app and assuring that the events are not lost. However, it can now report errors to a Snowplow collector as `diagnostic_error` events.
```

To activate this feature you only need to enable `trackerDiagnostic` ([#343][343]):

{% highlight java %}
TrackerBuilder trackerBuilder =
    new TrackerBuilder(emitter, namespace, appId, appContext)
        .trackerDiagnostic(true)
        ...
        .build();
Tracker.init(trackerBuilder);
{% endhighlight %}

> It will automatically report any internal tracker error. These will be fundamental for the troubleshooting of unexpected behaviour of the tracker.

Reviewer 3: ```suggestion
It will automatically report any internal tracker error to the configured Snowplow collector. These will be fundamental for the troubleshooting unexpected behaviour of the tracker.
```

> The sequence of those events can be sent to the Snowplow support team for analysis in order to spot tracker issues or the source of unexpected behaviour in advance.

Reviewer 3: ```suggestion
For Snowplow Insights customers, the sequence of tracker diagnostic events can be sent to the Snowplow support team for analysis in order to spot tracker issues or the source of unexpected behaviour.
```

“unstruct_event_com_snowplowanalytics_snowplow_diagnostic_error_1”: {
      “className”: “Tracker”,
      “message”: “pool-1-thread-2|Error tracker message”
}

<h2 id="iab">3. Fix for IAB enrichment filtering</h2>

As reported [here](https://discourse.snowplowanalytics.com/t/warning-iab-enrichment-treats-android-tracker-events-as-spider-generated/2482) the IAB enrichment incorrectly treats events sent by the Android tracker as spider-generated. This is due to the fact that the default User Agent used by the Android tracker, which is the one of the underlying library used to make HTTP calls (okHttp), is part of the IAB blacklist.

> The new Android tracker set a different default User Agent that is not filtered out by the IAB enrichment ([#359][359]). It's always possible to customize the User Agent used by the Android tracker setting the User Agent of the tracker's Subject: `setUserAgent`.

Reviewer 3: ```suggestion
The new Android tracker sets a different default User Agent that is not filtered out by the IAB enrichment ([#359][359]). It's also possible to customize the User Agent by setting the User Agent of the tracker's Subject: `setUserAgent`.
```

{% highlight java %}
Subject subject =
	new Subject.SubjectBuilder()
        ...
        .setUserAgent("my-custom-user-agent")
        ...
        .build();
TrackerBuilder trackerBuilder =
    new TrackerBuilder(emitter, namespace, appId, appContext)
        ...
        .subject(subject)
        ...
        .build();
Tracker.init(trackerBuilder);
{% endhighlight %}

<h2 id="custompath">4. Make POST path of Emitter configurable</h2>

> This is a feature already introduced in the Obj-C tracker at the version 1.1.0, but never released on Android ([#319][319]).

Reviewer 3: ```suggestion
This is a feature already introduced in the Obj-C tracker as of version 1.1.0, and is now available in the Android tracker ([#319][319]).
```

With the new version there is a new `customPostPath` parameter that makes the Emitter configurable.

> It lets to specify a custom path for POST requests useful when the user wants to use a proxy for receiving the events.

Reviewer 3: ```suggestion
It allows specifying a custom path for POST requests, this is useful when the user wants to use a proxy for receiving the events.
```

{% highlight java %}
Emitter emitter =
    new Emitter.EmitterBuilder(uri, context)
        ...
        .method(HttpMethod.POST)
        .customPostPath("com.acme.company/tpx")
        ...
        .build();
TrackerBuilder trackerBuilder =
    new TrackerBuilder(emitter, namespace, appId, appContext)
        ...
        .build();
Tracker.init(trackerBuilder);
{% endhighlight %}

<h2 id="updates">5. Updates and bug fixes</h2>

- Bump target Android API to 29 ([#357][357])

> We updated the library and the demo app to be fully compatible with Android 10 (API 29). It doesn't introduce any incompatibility with old versions and minimum Android OS version supported still the Android 4.0 (API 14).

Reviewer 3: ```suggestion
We updated the library and the demo app to be fully compatible with Android 10 (API 29). It doesn't introduce any incompatibility with old versions and minimum Android OS version supported still Android 4.0 (API 14).
```

- Cannot unset user id from subject ([#353][353])

> This bug causes issues when a parameter not required is set to null. In particular it has been reported that the `uid` parameter couldn't be cleared from the Subject once it was set the first time. Now, setting it to `null` clears the parameter.

Reviewer 3: ```suggestion
This bug causes issues when a parameter that is not required is set to null. In particular it has been reported that the `uid` parameter couldn't be cleared from the Subject once it was set the first time. Now, setting it to `null` clears the parameter.
```

- The method `Util.getEventId` is deprecated

It is a utility function in the `Util` class able to generate random UUID. It has been renamed to `getUUIDString`.

> The method `getEventId` will be removed in the 2.0 release.

Reviewer 3: ```suggestion
The method `getEventId` will be removed in the future v2.0 release.
```
